### PR TITLE
Fix C balance and FPE errors

### LINKED ISSUE #6120: Reinstate carbon balance check `err_found` for FATES cases
It appears that an older commit which was avoiding ending FATES runs due to carbon imbalance was accidentally reintroduced with #3986 after it had been removed via #3990:

https://github.com/E3SM-Project/E3SM/blob/3fa659c8cb50895b982e1ed7857628195182cf8e/components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90#L288-L317

The `if (.not. use_fates) then` statement should be removed.  Note that the phosphorous and nitrogen checks that were introduced via #3990 as well appear to be intact.

cc @rgknox @bishtgautam

### LINKED ISSUE #6177: FPE in `PhenologyMod.F90` with `ERS_D.hcru_hcru.I20TRGSWCNPRDCTCBC.pm-cpu_gnu.elm-erosion`
Using Jan23 master and `ERS_D.hcru_hcru.I20TRGSWCNPRDCTCBC.pm-cpu_gnu.elm-erosion`

```
 99: Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.
 99: 
 99: Backtrace for this error:
 99: #0  0x147b49653dbf in ???
 99: #1  0x166cbc5 in cropplantdate
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/components/elm/src/biogeochem/PhenologyMod.F90:2563
 99: #2  0x169a017 in __phenologymod_MOD_phenology
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/components/elm/src/biogeochem/PhenologyMod.F90:306
 99: #3  0x1563138 in __ecosystemdynmod_MOD_ecosystemdynnoleaching2
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/components/elm/src/biogeochem/EcosystemDynMod.F90:611
 99: #4  0x694d4d in __elm_driver_MOD_elm_drv
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/components/elm/src/main/elm_driver.F90:1032
 99: #5  0x65d258 in __lnd_comp_mct_MOD_lnd_run_mct
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/components/elm/src/cpl/lnd_comp_mct.F90:617
 99: #6  0x48441a in __component_mod_MOD_component_run
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/driver-mct/main/component_mod.F90:734
 99: #7  0x467be1 in __cime_comp_mod_MOD_cime_run
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/driver-mct/main/cime_comp_mod.F90:2920
 99: #8  0x481785 in cime_driver
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/driver-mct/main/cime_driver.F90:153
 99: #9  0x4817e8 in main
 99:    at /global/cfs/cdirs/e3sm/ndk/repos/me37-jan23/driver-mct/main/cime_driver.F90:23

```

also error with DEBUG and intel

### LINKED ISSUE #6203: Carbon Imbalance with CN runs
Discovered that `GridCBalanceCheck` doesn't use absolute value when error checking so negative imbalances can propagate through the Land Model.  [Line is here](https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90#L961C14-L961C26)

Running `e3sm_land_developer` test suite after changing to absolute value shows tests with CN enabled all FAIL with small ( ~`1.e-07`) imbalances.  Similar but independent error to keep in mind is Issue #6052

### PR DESCRIPTION
**Pull request by peterdschwartz**
Carbon balance was never checked correctly and let through negative values. 
Initialization of Ecosystem variables needed to be re-adjusted, and balance checking for FATES needed to be changed, which causes round-off differences in `CMASS_BALANCE_ERROR` for `fate_cold_allvars` test.

Also included divide-by-zero check in PhenologyMod.F90 to fix fpe in debug mode.

[non-BFB] for one FATES test.
Fixes #6120 
Fixes #6203 
Fixes #6177

### CODE REVIEW

**Comment by peterdschwartz** [comment]
@glemieux Here is a PR to fix the balance checking issues.  I did get DIFF for `fates_cold_allvars` but only for `CMASS_BALANCE_CHECK` but the values were still `~1E-16`.  

Let me know if how I implemented the new balance checking makes sense to you.

**Review by glemieux — APPROVED** [review]
The implementation looks good to me.  Thanks!

**Comment by peterdschwartz** [comment]
After doing a test merge I obtained FILLDIFF for the `fates_cold_allvars`. That I need to look into today

**Comment by rljacob** [comment]
What is the diff with?  A baseline from the previous code or is it a diff within a test like an ERS test?

**Comment by peterdschwartz** [comment]
Not baseline diff (though that is also there but makes sense to me).  It fails the COMP_base_rest portion of ERP test (same thing with the ERS version I made for debugging).

**Comment by rljacob** [comment]
Try a REP test.

**Comment by peterdschwartz** [comment]
REP passes just fine. I'm going to double check that each variable for carbon balance is read/written to the restart file, and maybe a longer simulation will reveal diff's accruing in other fields.

File: components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90 [hunk 1]
```diff
@@ -624,7 +621,7 @@ subroutine ColPBalanceCheck(bounds, &
       ! set time steps
       dt = dtime_mod
       kyr = year_curr; kmo = mon_curr; kda = day_curr; mcsec = secs_curr;
-
+      !nstep = nstep_mod
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90:624]: Please delete this line.

File: components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90 [hunk 2]
```diff
@@ -901,7 +898,14 @@ subroutine GridCBalanceCheck(bounds, col_cs, col_cf, grc_cs, grc_cf)
...
+         grc_coutputs              => grc_cf%coutputs              , & ! Output: [real(r8) (:) ] (gC/m2/s) column-level C outputs
+         beg_totpftc               =>  grc_cs%beg_totpftc          , & ! Input: [real(r8) (:)] (gC/m2) patch-level carbon aggregated to column level, incl veg and cpool
+         beg_cwdc                  =>  grc_cs%beg_cwdc             , & ! Input: [real(r8) (:)] (gC/m2) total column coarse woody debris carbon
+         beg_totsomc               =>  grc_cs%beg_totsomc          , & ! Input: [real(r8) (:)] (gC/m2) total column soil organic matter carbon
+         beg_totlitc               =>  grc_cs%beg_totlitc          , & ! Input: [real(r8) (:)] (gC/m2) total column litter carbon
+         beg_totprodc              =>  grc_cs%beg_totprodc         , & ! Input: [real(r8) (:)] (gC/m2) total column wood product carbon
+         beg_ctrunc                =>  grc_cs%beg_ctrunc           , & ! Input: [real(r8) (:)] (gC/m2) total column truncation carbon sink
+         beg_cropseedc_deficit     =>  grc_cs%beg_cropseedc_deficit  & ! Input: [real(r8) (:)] (gC/m2) column carbon pool for seeding new growth
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90:908]: Please remove an extra whitespace after `=>  grc_cs%` to be `=> grc_cs%`

File: components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90 [hunk 3]
```diff
@@ -940,13 +944,21 @@ subroutine GridCBalanceCheck(bounds, col_cs, col_cf, grc_cs, grc_cf)
       call c2g(bounds, col_som_c_leached(bounds%begc:bounds%endc), grc_som_c_leached(bounds%begg:bounds%endg), &
                c2l_scale_type = 'unity', l2g_scale_type = 'unity')
       call c2g(bounds, col_som_c_yield(bounds%begc:bounds%endc), grc_som_c_yield(bounds%begg:bounds%endg), &
-               c2l_scale_type = 'unity', l2g_scale_type = 'unity')
+               c2l_scale_type = 'unity', l2g_scale_type = 'unity')  
 
+      if(use_fates) then
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90:949]: please add a whitespace `if (use_fates)`

File: components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90 [hunk 4]
```diff
@@ -940,13 +944,21 @@ subroutine GridCBalanceCheck(bounds, col_cs, col_cf, grc_cs, grc_cf)
...
+               c2l_scale_type = 'unity', l2g_scale_type = 'unity')
+      end if 
+         
       dt = real( get_step_size(), r8 )
       nstep = get_nstep()
 
       do g = bounds%begg, bounds%endg
-         grc_cinputs(g) = grc_gpp(g) + grc_dwt_seedc_to_leaf(g) + grc_dwt_seedc_to_deadstem(g)
+
+         if(.not. use_fates) then
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90:959]: similarly add a whitespace.

File: components/elm/src/biogeochem/PhenologyMod.F90 [hunk 5]
```diff
@@ -2559,8 +2559,11 @@ subroutine CropPlantDate (num_soilp, filter_soilp, num_pcropp, filter_pcropp, &
          call calculate_eto(t_ref2m(p), netrad, eflx_soil_grnd(p), forc_pbot(t), forc_rh(t), forc_wind(t), dt, ETout)
          ! monthly ETo
          ETo(p,kmo) = ETo(p,kmo) + ETout
+         
          ! calculate the P:PET for each month
-         p2ETo(p,kmo) = xp(p,kmo)/ETo(p,kmo)
+         if( abs(ETo(p,kmo)) > 0._r8) then
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/PhenologyMod.F90:2564]: - please add whitespace
- Shouldn't there also be code corresponding to an `else` condition?

File: components/elm/src/data_types/ColumnDataType.F90 [hunk 6]
```diff
@@ -7764,6 +7766,8 @@ subroutine col_cf_setvalues ( this, num_column, filter_column, value_column)
        this%litterc_loss(i)              = value_column
        
        this%nee(i)                       = value_column
+       this%er(i)                        = value_column !! Grid Balance Error 
+       this%som_c_leached(i)             = value_column !! Grid Balance Error
```
Comment (bishtgautam) [inline components/elm/src/data_types/ColumnDataType.F90:7770]: Can you please remove the `!! Grid Balance Error`

**Review by bishtgautam — CHANGES_REQUESTED** [review]
I have minor stylistic code change requests for this PR.

**Comment by rljacob** [comment]
@peterdschwartz what is the status of this PR?

**Comment by peterdschwartz** [comment]
@rljacob I met with @glemieux  a couple of weeks ago. He mentioned a possible time-step dependent part of FATES that could be the cause, but I think he's been busy with the FATES api update and our follow up meeting with @thorntonpe was cancelled due to lab holiday.  

Today, I'll try increasing the history output to be every timestep and see if that provides any useful information.

**Comment by peterdschwartz** [comment]
Update:

Increasing output freq finally showed a new diff in a field.  Looking into it further, it appears that the tendency term for carbon are not added to the leaching term due to some issue 5 years ago (according to git blame atleast), but it is for nitrogen and phosphorus. 

Hopefully the straightforward solution works here.
```
 LITR2C_TNDNCY_VERT_TRANS   (lndgrid,levdcmp,time)  t_index =    194   194
          3     5760  (   153,     2,     1) (   153,     1,     1) (   237,     3,     1) (   247,    10,     1)
                3165   1.043116371057806E-09  -1.643725044964128E-09 2.2E-19  2.944499461962624E-12 8.1E-11  1.293930177203629E-16
                3165   1.043116371057806E-09  -1.643725044964128E-09          2.944499678803059E-12          1.293930309552527E-16
                5760  (   153,     2,     1) (   153,     1,     1)
          avg abs field values:    4.034842251976478E-11    rms diff: 0.0E+00   avg rel diff(npos):  8.1E-11
                                   4.034842251976478E-11                        avg decimal digits(ndif):  7.1 worst:  7.0
 RMS LITR2C_TNDNCY_VERT_TRANS         0.0000E+00            NORMALIZED  0.0000E+00
```

**Comment by glemieux** [comment]
@peterdschwartz can you point me to the line of code with the git blame that you noted earlier, please?

**Comment by peterdschwartz** [comment]
@glemieux This [Code block](https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/data_types/ColumnDataType.F90#L7531) is what I was referring to. The transport tendency is not added to carbon leached pool even though the corresponding terms are added for Nitrogen and Phosphorus.  

More experimenting showed that bypassing the SoilLittVertTransp calculation, the restart failure goes away. For FATES, the `UpdateLitterFluxes`  is done just prior that adjusts the inputs to the `SoilLittVertTransp` that could be the cause.  [Line](https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeochem/EcosystemDynMod.F90#L706)

**Comment by peterdschwartz** [comment]
Since there's a previous fates related issue linked above with the same errors, I'm going to just clean the branch up for the other issues and follow the review comments. The other issue can be used to the issue.

**Comment by glemieux** [comment]
> Since there's a previous fates related issue linked above with the same errors, I'm going to just clean the branch up for the other issues and follow the review comments. The other issue can be used to the issue.

Sorry for the delay in trying to address the fates-side of this.  Just to clarify, you're saying your going to move forwards towards integration @peterdschwartz?  If so, that sounds good to me.

**Comment by peterdschwartz** [comment]
> > Since there's a previous fates related issue linked above with the same errors, I'm going to just clean the branch up for the other issues and follow the review comments. The other issue can be used to the issue.
> 
> Sorry for the delay in trying to address the fates-side of this. Just to clarify, you're saying your going to move forwards towards integration @peterdschwartz? If so, that sounds good to me.

@glemieux  Yes, and now I know how to how to reliably reproduce the issue, we can use #6125 to track progress. I will share a branch with the changes to do it.

@bishtgautam I have rebased the branch and addressed your comments.  There's potentially a couple of different ways to handle the `else` condition in the PhenologyMod, so I went with what I believe to be the simplest and added a comment explaining why.  Let me know what you think

e3sm_developer came back [BfB] with the exception of one FATES test that will FAIL restart comparison for the time being.

File: components/elm/src/biogeochem/PhenologyMod.F90 [hunk 7]
```diff
@@ -2559,8 +2559,17 @@ subroutine CropPlantDate (num_soilp, filter_soilp, num_pcropp, filter_pcropp, &
...
-         p2ETo(p,kmo) = xp(p,kmo)/ETo(p,kmo)
+         if ( abs(ETo(p,kmo)) > 0._r8) then 
+            p2ETo(p,kmo) = xp(p,kmo)/ETo(p,kmo)
+         else ! P:PET is undefined.
+            ! Setting to a fill value ( 'spval' ) would
+            ! require nested if statements due to 
+            ! the weighting of previous years (i.e., p2ETo and prev_p2ETo_bar )
+            ! So, set to zero for simplicity.
+            p2ETo(p,kmo) = 0._r8
+         end if
```
Comment (bishtgautam) [inline components/elm/src/biogeochem/PhenologyMod.F90:2572]: @peterdschwartz This looks good to me. Thanks for the comment explaining the logic.

**Comment by rljacob** [comment]
This is ready to merge.

**Review by bishtgautam — APPROVED** [review]

**Pull request merged**